lib/capability/capability.go

```diff
@@ -0,0 +1,94 @@
+package capability
+
+import (
+	"fmt"
+	"strings"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+const CapabilityAnnotation = "capability.openshift.io/name"
+
+type ClusterCapabilities struct {
+	KnownCapabilities map[configv1.ClusterVersionCapability]struct{}
+
+	// value indicates whether capability was explicitly included
+	EnabledCapabilities map[configv1.ClusterVersionCapability]bool
+}
+
+func SetCapabilities(config *configv1.ClusterVersion) ClusterCapabilities {
+	var capabilities ClusterCapabilities
+	capabilities.KnownCapabilities = setKnownCapabilities(config)
+	capabilities.EnabledCapabilities = setEnabledCapabilities(config)
+	return capabilities
+}
+
+func GetCapabilitiesStatus(capabilities ClusterCapabilities) configv1.ClusterVersionCapabilitiesStatus {
+	var status configv1.ClusterVersionCapabilitiesStatus
+	for k, _ := range capabilities.EnabledCapabilities {
+		status.EnabledCapabilities = append(status.EnabledCapabilities, k)
+	}
+	for k, _ := range capabilities.KnownCapabilities {
+		status.KnownCapabilities = append(status.KnownCapabilities, k)
+	}
+	return status
+}
+
+func CheckResourceEnablement(annotations map[string]string, capabilities ClusterCapabilities) error {
+	val, ok := annotations[CapabilityAnnotation]
+	if ok {
+		caps := strings.Split(val, "+")
+		numCaps := len(caps)
+		unknownCaps := make([]string, numCaps)
+		disabledCaps := make([]string, numCaps)
+
+		for _, c := range caps {
+			trimmedCap := strings.TrimSpace(c)
+
+			if _, ok = capabilities.KnownCapabilities[configv1.ClusterVersionCapability(trimmedCap)]; !ok {
+				unknownCaps = append(unknownCaps, trimmedCap)
+			} else if _, ok = capabilities.EnabledCapabilities[configv1.ClusterVersionCapability(trimmedCap)]; !ok {
+				disabledCaps = append(disabledCaps, trimmedCap)
+			}
+		}
+		if len(unknownCaps) > 0 {
+			return fmt.Errorf("unrecognized capability names: %s", strings.Join(unknownCaps, ", "))
+		}
+		if len(disabledCaps) > 0 {
+			return fmt.Errorf("disabled capabilities: %s", strings.Join(disabledCaps, ", "))
+		}
+	}
+	return nil
+}
+
+func setKnownCapabilities(config *configv1.ClusterVersion) map[configv1.ClusterVersionCapability]struct{} {
+	known := make(map[configv1.ClusterVersionCapability]struct{})
+
+	for _, v := range configv1.ClusterVersionCapabilitySets {
+		for _, capability := range v {
+			if _, ok := known[capability]; ok {
+				continue
+			}
+			known[capability] = struct{}{}
+			config.Status.Capabilities.KnownCapabilities = append(config.Status.Capabilities.KnownCapabilities, capability)
```

Comment:
Do we have reliable locking on this?  With the status sync'er running in parallel, I'd expect at least something like building a slice in a local variable, and after completing the local array, slotting it in with:

```go
config.Status.Capabilities.KnownCapabilities = knownCapabilities
```

Reply:
This is old code and will be removed but the question of the need for a lock is valid.

The capabilities reside in get set `SyncWork` and are set in `Update` so are protected by the lock there. Just yesterday I added setting `CapabilitiesStatus` in `SyncWorkerStatus` also in `Update` [here](https://github.com/openshift/cluster-version-operator/blob/abafa6061c661debb0c37b07fab039c827f2df0c/pkg/cvo/sync_worker.go#L436) so it's available to `syncStatus` [here](https://github.com/openshift/cluster-version-operator/blob/abafa6061c661debb0c37b07fab039c827f2df0c/pkg/cvo/status.go#L204).